File: solutions/098.Validate_Binary_Search_Tree/AC_dfs_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.java
*  Create Date: 2015-03-03 14:57:45
*  Descripton:   
*/

import java.util.Scanner;

public class Solution {

    // Definition for binary tree
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    private boolean dfs(TreeNode root, long minval, long maxval) {
        if (root == null)
            return true;
        return root.val > minval && root.val < maxval &&
                dfs(root.left, minval, root.val) && dfs(root.right, root.val, maxval);
    }

    public boolean isValidBST(TreeNode root) {
        return dfs(root, (long)Integer.MIN_VALUE - 1, (long)Integer.MAX_VALUE + 1);
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 1};
        System.out.println("no case");
    }
}

File: solutions/106.Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/AC_dfs_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_n.cpp
*  Create Date: 2015-03-09 10:07:27
*  Descripton:  Just simulation.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    TreeNode *dfs(vector<int> &inorder, vector<int> &postorder, int is, int ps, int len) {
        if (len <= 0)
            return 0;
        TreeNode *root = new TreeNode(postorder[ps]);
        for (int i = 0; i < len; ++i) {
            if (inorder[is - i] == root->val) {
                root->right = dfs(inorder, postorder, is, ps - 1, i);
                root->left = dfs(inorder, postorder, is - i - 1, ps - i - 1, len - i - 1);
                break;
            }
        }
        return root;
    }
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        return dfs(inorder, postorder, inorder.size() - 1, postorder.size() - 1, inorder.size());
    }
};

int main() {

	return 0;
}

